### DISCUSSION TITLE
Why so slow when I execute 'insert' into an encrypted database？

### DISCUSSION THREAD

**Discussion by 724198847** [rusqlite/rusqlite#1148] [Questions]
```toml
[dependencies]
rusqlite = {version = "0.27.0", features = ["sqlcipher"]}
```

```rs
extern crate rusqlite;
use rusqlite::{Connection};
use std::time::{SystemTime, UNIX_EPOCH};

fn main() {
    let new_uri = "./test.sqlite";
    let conn = rusqlite::Connection::open(new_uri);
    match conn {
        Ok(conn) => {
            let res_11 = conn.execute("PRAGMA key = '3e30431105262861ffad1d2ada34e82c';", []);
            let res_12 = conn.execute("PRAGMA cipher_compatibility = '4';", []);
            // let res_1 = conn.execute("CREATE TABLE test_table (id, name)", []);

            let start = SystemTime::now();
            // let mut i = 2;
            // while i < 100000 {
                let res_3 = conn.execute("INSERT INTO test_table VALUES ('2', '今天')", []);
            //     i += 1;
            // }
            
            let time_consuming = SystemTime::now().duration_since(start).unwrap_or_default();
            println!("[Rust][sqlite][execute] it takes {:?} ", time_consuming);
        }
        Err(err) => println!("{}", err)
    }
}
```

This is my demo's code and dependencies. When the amount of data is small, the execution is relatively fast, within ten milliseconds. But when there are tens of thousands of data, it takes nearly 200 milliseconds to execute an insertion, which is very abnormal. Please help, am I missing some settings？（It takes within ten milliseconds in DB Browser）

**Comment by thomcc** [comment]
Yeah, there are a few settings. See https://github.com/rusqlite/rusqlite/discussions/1119#discussioncomment-2107571 (the whole thread, you have to click to expand it) for discussion of some of them. For example, you should consider using WAL (and if you use WAL, you can turn `PRAGMA synchronous=normal`).

But for SQLcipher *specifically*, try `PRAGMA cipher_memory_security = false`. This controls whether or not `mlock` is called for every malloc call and `munlock` is called for every free. Disabling this means you can't prevent stuff like cold boot attacks, but these are very very hard to prevent in rust anyway (since even stuff like `zeroize` can't help with value types on the stack being moved).

When I worked at mozilla, we did this in the mobile firefoxes: https://github.com/mozilla/application-services/blob/0d4ae3876b1f20da2b40c6fe5701863d0b6696dd/components/places/src/db/db.rs#L57-L60. As noted there, it helps a lot for insert-heavy workloads.

**Reply by thomcc** [reply]
Here are the docs for it: https://www.zetetic.net/sqlcipher/sqlcipher-api/#cipher_memory_security. Note that despite what it says, it also controls the use of `mlock`/`munlock` (`VirtualLock`/`VirtualUnlock` on windows), which are the primary cause of the slowdown, rather than the wiping in particular. These operations prevent the memory of the DB from every being swapped to disk (when the program is in the background for example).

That said, encryption also does cause a perf hit, but not as much as this.

Also, be sure you're running with `cargo run --release`. Debug builds will be slow (but will compile faster).

**Reply by 724198847** [reply]
Thanks!!!
